### ISSUE TITLE
Problem with compound locations wrapping origin in SeqIO

### ISSUE THREAD

**Issue by mmcguffi** [biopython/biopython#2247]
### Setup

I am reporting a problem with Biopython version, Python version, and operating
system as follows:

3.7.4 (default, Jul  9 2019, 18:13:23) 
[Clang 10.0.1 (clang-1001.0.46.4)]
CPython
Darwin-18.2.0-x86_64-i386-64bit
1.74

### Expected behaviour
When opening a gbk file that has a feature which contains a(n incorrectly) specified location which wraps the origin, but also initially contains a compound location, this compound location should be "repaired" and also throw a warning.

### Actual behaviour

When trying to open a .gbk that has this issue, this error is thrown:

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-4e2d1f4e6ad8> in <module>()
     12         file.write("%s\n" % line)
     13 with open(tmp.name,'r') as file_handle:
---> 14     record_dict = SeqIO.to_dict(SeqIO.parse(file_handle, 'gb'))
     15 tmp.close()

/usr/local/lib/python3.7/site-packages/Bio/SeqIO/__init__.py in to_dict(sequences, key_function)
    808 
    809     d = _dict()
--> 810     for record in sequences:
    811         key = key_function(record)
    812         if key in d:

/usr/local/lib/python3.7/site-packages/Bio/SeqIO/__init__.py in parse(handle, format, alphabet)
    659             raise ValueError("Unknown format '%s'" % format)
    660         # This imposes some overhead... wait until we drop Python 2.4 to fix it
--> 661         for r in i:
    662             yield r
    663 

/usr/local/lib/python3.7/site-packages/Bio/GenBank/Scanner.py in parse_records(self, handle, do_features)
    491         # This is a generator function
    492         while True:
--> 493             record = self.parse(handle, do_features)
    494             if record is None:
    495                 break

/usr/local/lib/python3.7/site-packages/Bio/GenBank/Scanner.py in parse(self, handle, do_features)
    475                                     feature_cleaner=FeatureValueCleaner())
    476 
--> 477         if self.feed(handle, consumer, do_features):
    478             return consumer.data
    479         else:

/usr/local/lib/python3.7/site-packages/Bio/GenBank/Scanner.py in feed(self, handle, consumer, do_features)
    447         # Features (common to both EMBL and GenBank):
    448         if do_features:
--> 449             self._feed_feature_table(consumer, self.parse_features(skip=False))
    450         else:
    451             self.parse_features(skip=True)  # ignore the data

/usr/local/lib/python3.7/site-packages/Bio/GenBank/Scanner.py in _feed_feature_table(consumer, feature_tuples)
    397         for feature_key, location_string, qualifiers in feature_tuples:
    398             consumer.feature_key(feature_key)
--> 399             consumer.location(location_string)
    400             for q_key, q_value in qualifiers:
    401                 if q_value is None:

/usr/local/lib/python3.7/site-packages/Bio/GenBank/__init__.py in location(self, content)
   1118                 locs.append(SeqFeature.FeatureLocation(int(s) - 1,
   1119                                                        int(e),
-> 1120                                                        strand))
   1121             if len(locs) < 2:
   1122                 # The CompoundLocation will raise a ValueError here!

/usr/local/lib/python3.7/site-packages/Bio/SeqFeature.py in __init__(self, start, end, strand, ref, ref_db)
    759             raise ValueError('End location ({}) must be greater than or equal '
    760                              'to start location ({})'.format(self.end,
--> 761                                                              self.start))
    762         self.strand = strand
    763         self.ref = ref

ValueError: End location (450) must be greater than or equal to start location (9712)
```
It appears that Snapgene never fixed this issue on their end as far as I can tell.

### Steps to reproduce

```python
import requests
from tempfile import NamedTemporaryFile
from Bio import SeqIO

url='https://media.addgene.org/snapgene-media/v1.6.2-0-g4b4ed87/sequences/77/92/177792/addgene-plasmid-85619-sequence-177792.gbk'
response = requests.get(url)
urldata = response.text

tmp = NamedTemporaryFile()
with open(tmp.name, "w") as file:
    for line in urldata.split("\r\n"):
        file.write("%s\n" % line)
with open(tmp.name,'r') as file_handle:
    record_dict = SeqIO.to_dict(SeqIO.parse(file_handle, 'gb'))
tmp.close()
```

**Comment by peterjc** [comment]
Thank you for sharing a self-contained example. I could reproduce this with the latest Biopython code in github. Shorter example code using this file,

```
$ python -c "from Bio import SeqIO; print(list(SeqIO.parse('addgene-plasmid-85619-sequence-177792.gbk', 'gb')))"
...
ValueError: End location (450) must be greater than or equal to start location (9712)
```

This is from the final feature in the table, a CDS with a compound location (a join):

```
LOCUS       Exported                9877 bp ds-DNA     circular SYN 06-AUG-2018
DEFINITION  Expresses V5 tagged mutant yki in Drosophila..
...
     CDS             join(7455..7526,7928..8201,8276..8930,8992..9307,9511..9642,
                     9713..450)
                     /codon_start=1
                     /gene="white"
                     /product="Drosophila white gene eye color pigment"
                     /label=mini-white
                     /note="This is a modified version of the white gene lacking
                     part of the first intron."
...
```

The problem is the last exon, ``9713..450``, which ought to be split as ``9713..9877,1..450`` since it spans the origin.

Since the original special case code (#1770 added in Biopython 1.73), we have already extended the special case code for this once to handle fuzzy locations (#2162 added in Biopython 1.74), so you want to extend it again to handle compound locations?

I maintain that the best fix is to get the tools producing these invalid files corrected (and being practical, fixing the files manually as and where you find one in your own datasets), but am willing to review yet another special case with tests if it isn't too invasive.

How often have you seen these? Is it really common enough to justify automating.

If you are a paying customer of SnapGene, please to raise this with their support team.

**Comment by nimne** [comment]
I'm happy to take a look at this to see if the fix is simple. I hadn't considered introns when I made my last fix!

Fortunately, it looks like snapgene (4.2.1) fixes these files - it is able to open them, and saves them with the correct position: ```join(7455..7526,7928..8201,8276..8930,8992..9307,9511..9642,9713..9877,1..450)```

While I'm looking at this, are the other cases to be concerned with? Are inexact positions valid in a compound location like this? For example:
```join(8992..9307,9511..9642,<9713..450)```

**Comment by peterjc** [comment]
@nimne I think fuzziness at the very start or end is more likely, but yes, that kind of thing is possible too in principle.

@mmcguffi what version of SnapGene are you using?

**Comment by mmcguffi** [comment]
@peterjc I don't actually use SnapGene, I am doing some computations with data from addgene, which uses SnapGene to annotate their plasmids -- I suppose that the annotations are locked-in from an older version. @nimne I looked at ~30,000 plasmids and this is was the only situation that I saw that threw an error, and there were only a handful at that.

**Comment by peterjc** [comment]
Belatedly closing, should have been fixed with #2254 (parsed with a warning)
